### PR TITLE
908 workflows: first step restores from source_data

### PR DESCRIPTION
**Pull request by szymonlopaciuk**
## Description
Add a step to the beginning of the article and author workflows that restores it's contents (`data` & `extra_data`) from `source_data`.

## Related Issue
https://its.cern.ch/jira/browse/INSPIR-908

## Motivation and Context
This is a step towards restartable workflows.

## Checklist:

- [x] I have all the information that I need (if not, move to `RFC` and look for it).
- [x] I linked the related issue(s) in the corresponding commit logs.
- [x] I wrote [good commit log messages](https://github.com/torvalds/subsurface-for-dirk/blob/5f15ad5a86ada3c5e574041a5f9d85235322dabb/README#L92-L119).
- [x] My code follows the code style of this project.
- [x] I've added any new docs if API/utils methods were added.
- [x] I have updated the existing documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by ammirate** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

File: tests/integration/workflows/test_arxiv_merge.py [hunk 1]
```diff
@@ -167,8 +168,8 @@ def test_merge_with_disabled_merge_on_update_feature_flag(
         factory = TestRecordMetadata.create_from_file(
             __name__, 'merge_record_arxiv.json', index_name='records-hep')
 
-        record_update = RECORD_WITHOUT_CONFLICTS
-        eng_uuid = start('article', [record_update])
+        record_update = build_workflow(RECORD_WITHOUT_CONFLICTS)
```
Comment (david-caro) [inline tests/integration/workflows/test_arxiv_merge.py]: record_update_id = build_workflow().id

File: inspirehep/modules/workflows/tasks/actions.py [hunk 2]
```diff
@@ -615,3 +615,13 @@ def jlab_ticket_needed(obj, eng):
...
+
+
+@with_debug_logging
+def load_from_source_data(obj, enj):
+    """Restore the workflow data and extra_data from source_data."""
+    source_data = obj.extra_data['source_data']
+    obj.data = source_data['data']
+    obj.extra_data = source_data['extra_data']
+    obj.extra_data['source_data'] = deepcopy(source_data)
+    obj.save()
```
Comment (david-caro) [inline inspirehep/modules/workflows/tasks/actions.py]: Throw exception with nice message pointing to dev work if there's no source_data (so the curators don't go crazy restarting the workflow)

Comment (david-caro) [reply inspirehep/modules/workflows/tasks/actions.py]: Copy over the _task_history until we have proper logs :cry:

File: tests/integration/workflows/test_arxiv_workflow.py [hunk 3]
```diff
@@ -958,4 +968,45 @@ def test_match_wf_in_error_goes_in_initial_state(workflow_app):
...
+        data_type='hep',
+        **broken_record
+    )
+    eng_uuid = start('article', object_id=record_workflow.id)
+
+    eng = WorkflowEngine.from_uuid(eng_uuid)
+    obj = eng.objects[0]
+
+    assert obj.status != ObjectStatus.ERROR
+    assert '_error_msg' not in obj.extra_data
```
Comment (david-caro) [inline tests/integration/workflows/test_arxiv_workflow.py]: For the tests, I would:
* e2e testing the feature (creating a workflow that halts, modifying it, restarting and making sure it restored the data and extra_data)
* integration:
** test that it throws the exception on no source_data
** test that it restores the extra_data and data for the rest (probably you might want a custom wokrflow)
*** For wflw with both extra_data and data
*** For wflw with empty extra_data
*** For worfklow with empty data
*** For wflw with both empty
** test that it does not destroy the task_history (check also if all workflows have task_history)

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

File: tests/integration/workflows/test_workflows_tasks_actions.py [hunk 4]
```diff
@@ -472,3 +479,79 @@ def test_count_reference_coreness(
...
+
+    del workflow_app.extensions['invenio-workflows'].workflows['load_source_data']
+
+
+def test_workflow_loads_from_source_data_fails_on_no_source_data(
+    load_from_source_data_workflow,
+    workflow_app,
+    record_from_db,
+):
+    workflow_id = workflow_object_class.create(data_type='hep', data=record_from_db).id
```
Comment (david-caro) [inline tests/integration/workflows/test_workflows_tasks_actions.py]:
```python
extra_data_without_source_data = {}
workflow_id = workflow_object_class.create(
    data_type='hep',
    data=record_from_db,
    extra_data=extra_data_without_source_data,
).id
```

File: tests/e2e/test_submissions.py [hunk 5]
```diff
@@ -58,58 +58,111 @@ def _workflows_in_status(holdingpen_client, num_entries, status):
...
-
-    assert entry.display_name == 'Homer Simpson'
+    workflow_id = halted_entry.workflow_id
+    entry = inspire_client.holdingpen.get_detail_entry(workflow_id)
 
+    # Assert that source_data is there
     result_source_data = entry.extra_data['source_data']
 
     assert result_source_data['data']
     assert result_source_data['extra_data']
```
Comment (david-caro) [inline tests/e2e/test_submissions.py]: Remove the checks for extra_data/source_data and data stuff up here (that's internal stuff)

Comment (david-caro) [reply tests/e2e/test_submissions.py]: Then we can remove the extra_data key from the HoldingpenEntry object thingie

File: tests/integration/workflows/test_workflows_tasks_actions.py [hunk 6]
```diff
@@ -472,3 +479,79 @@ def test_count_reference_coreness(
...
+
+    start('load_source_data', object_id=workflow_id)
+
+    workflow_object = workflow_object_class.get(workflow_id)
+
+    assert workflow_object.data['titles'][0]['title'] == 'Fancy title for a new record changed!'
+
+    workflow_object.callback_pos = [0]
+    workflow_object.save()
+    db.session.commit()
```
Comment (david-caro) [inline tests/integration/workflows/test_workflows_tasks_actions.py]: Remove the mess_up step, and manually modify the workflow, then restart it, and see that it got reverted.

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by szymonlopaciuk** [comment]
ci test workflows please

**Comment by szymonlopaciuk** [comment]
ci test please workflows

**Comment by szymonlopaciuk** [comment]
ci test please

**Comment by ammirate** [comment]
ci test workflow please

**Comment by ammirate** [comment]
ci test please workflows

**Comment by ammirate** [comment]
ci merge please

**Comment by szymonlopaciuk** [comment]
ci merge please

**Pull request merged**